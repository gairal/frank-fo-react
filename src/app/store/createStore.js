import { applyMiddleware, compose, createStore } from 'redux';
import thunk from 'redux-thunk';
import { browserHistory } from 'react-router';
import reducers from './reducers';
import { updateLocation } from './location';

export default (initialState = {}) => {
  // ======================================================
  // Middleware Configuration
  // ======================================================
  const middleware = [thunk];

  // ======================================================
  // Store Instantiation and HMR Setup
  // ======================================================
  const store = createStore(
    reducers(),
    initialState,
    compose(applyMiddleware(...middleware)),
  );
  store.asyncReducers = {};

  // To unsubscribe, invoke `store.unsubscribeHistory()` anytime
  store.unsubscribeHistory = browserHistory.listen(updateLocation(store));

  return store;
};
