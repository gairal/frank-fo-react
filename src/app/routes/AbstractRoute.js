import { connect } from 'react-redux';
import { injectReducer } from '@/store/reducers';
import * as actions from '../containers/Loader';

export default class AbstractRoute {
  constructor(store, key, icon, path = '') {
    this.store = store;
    this.key = key;
    this.icon = icon;

    this.API_URL_BASE = 'https://api-dot-com-gairal-frank.appspot.com';
    this.API_URL = `${this.API_URL_BASE}/${this.key}/${path}`;

    this.initialState = {
      data: [],
    };

    this.ACTIONS = {};
    this.ACTION_HANDLERS = {};

    this.mapStateToProps = {
      [this.key]: 'data',
    };
    this.mapDispatchToProps = {};
  }

  get connected() {
    const mapStateToProps = state =>
      Object.keys(this.mapStateToProps).reduce(
        (acc, e) => ({
          ...acc,
          [e]: state[this.key][this.mapStateToProps[e]],
        }),
        {},
      );

    return connect(
      mapStateToProps,
      this.mapDispatchToProps,
    )(this.component);
  }

  get route() {
    injectReducer(this.store, { key: this.key, reducer: this.getReducer() });
    return {
      key: this.key,
      icon: this.icon,
      path: `/${this.key}`,
      component: this.connected,
    };
  }

  getReducer() {
    const { initialState, ACTION_HANDLERS } = this;

    return (state = initialState, action) => {
      const handler = ACTION_HANDLERS[action.type];
      return handler ? handler(state, action) : state;
    };
  }

  // reducer functions
  static fetchSuccess(state, action) {
    return {
      ...state,
      data: action.payload.json,
    };
  }

  static fetchFail(state) {
    return state;
  }

  loadFactory() {
    const { API_URL } = this;
    const { success, fail } = this.constructor;

    return () => dispatch => {
      dispatch(actions.showLoader());

      const fetchOptions = {
        method: 'GET',
      };

      return fetch(API_URL, fetchOptions)
        .then(response => response.json())
        .then(json => {
          dispatch(actions.hideLoader());
          return dispatch(success(json));
        })
        .catch(err => {
          dispatch(actions.hideLoader());
          return dispatch(fail(err));
        });
    };
  }

  init() {
    this.ACTION_HANDLERS[this.ACTIONS.FETCH_SUCCESS] =
      AbstractRoute.fetchSuccess;
    this.ACTION_HANDLERS[this.ACTIONS.FETCH_FAILURE] = AbstractRoute.fetchFail;

    this.mapDispatchToProps.load = this.loadFactory();
  }

  /**
   * Default export method
   *
   * @static
   * @param {*} store
   * @memberof AbstractRoute
   */
  static export(store) {
    return new this(store).route;
  }
}
