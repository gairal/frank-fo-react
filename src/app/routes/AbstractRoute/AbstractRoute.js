import { connect } from 'react-redux';
import { injectReducer } from '../../store/reducers';

export default class AbstractRoute {
  constructor(store, key, icon, path = '') {
    this.store = store;
    this.key = key;
    this.icon = icon;

    this.API_URL = `https://com-gairal-frank-bo.appspot.com/${this.key}/${path}`;

    this.initialState = {
      data: [],
    };

    this.ACTIONS = {};
    this.ACTION_HANDLERS = {};

    this.mapStateToProps = {
      [this.key]: 'data',
    };
    this.mapDispatchToProps = {};
  }

  get connected() {
    const mapStateToProps = state => Object.keys(this.mapStateToProps)
      .reduce((acc, e) => ({
        ...acc,
        [e]: state[this.key][this.mapStateToProps[e]],
      }), {});

    return connect(mapStateToProps, this.mapDispatchToProps)(this.component);
  }

  get route() {
    injectReducer(this.store, { key: this.key, reducer: this.getReducer() });
    return {
      key: this.key,
      icon: this.icon,
      path: `/${this.key}`,
      component: this.connected,
    };
  }

  getReducer() {
    const { initialState, ACTION_HANDLERS } = this;

    return (state = initialState, action) => {
      const handler = ACTION_HANDLERS[action.type];
      return handler ? handler(state, action) : state;
    };
  }

  // reducer functions
  static fetchSuccess(state, action) {
    return ({
      ...state,
      data: action.payload.json,
    });
  }

  static fetchFail(state) {
    return state;
  }

  static showLoader() {
    return {
      type: 'LOADER_SHOW',
    };
  }

  static hideLoader() {
    return {
      type: 'LOADER_HIDE',
    };
  }

  loadFactory() {
    const { API_URL } = this;
    const { success, fail } = this.constructor;

    return () => (dispatch) => {
      dispatch(AbstractRoute.showLoader());

      const fetchOptions = {
        method: 'GET',
      };

      return fetch(API_URL, fetchOptions)
        .then(response => response.json())
        .then((json) => {
          dispatch(AbstractRoute.hideLoader());
          return dispatch(success(json));
        })
        .catch((err) => {
          dispatch(AbstractRoute.hideLoader());
          return dispatch(fail(err));
        });
    };
  }

  init() {
    this.ACTION_HANDLERS[this.ACTIONS.FETCH_SUCCESS] = AbstractRoute.fetchSuccess;
    this.ACTION_HANDLERS[this.ACTIONS.FETCH_FAILURE] = AbstractRoute.fetchFail;

    this.mapDispatchToProps.load = this.loadFactory();
  }
}
