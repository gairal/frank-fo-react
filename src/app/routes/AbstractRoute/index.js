import AbstractRoute from './AbstractRoute';

export default AbstractRoute;
