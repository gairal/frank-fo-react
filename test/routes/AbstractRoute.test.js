import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import createStore from '@/store/createStore';
import AbstractRoute from '@/routes/AbstractRoute';

describe('routes > AbstractRoute', () => {
  const store = createStore();
  const route = new AbstractRoute(store, 'test', 'test', 'test');

  route.component = () => (<div />);
  route.ACTIONS = {
    FETCH_SUCCESS: 'TEST_FETCH_SUCCESS',
    FETCH_FAILURE: 'TEST_FETCH_FAILURE',
  };
  route.init();

  AbstractRoute.success = jest.fn();
  AbstractRoute.fail = jest.fn();

  const connectedRoute = route.route;
  const Connected = connectedRoute.component;
  const component = renderer.create(<Provider store={store}><Connected /></Provider>);

  it('Has default props', () => {
    expect(route.key).toEqual('test');
    expect(route.icon).toEqual('test');
    expect(route.API_URL_BASE).toBeDefined();
    expect(route.API_URL).toBeDefined();
    expect(route.API_URL).toEqual(`${route.API_URL_BASE}/test/test`);
    expect(route.mapDispatchToProps.load).toBeDefined();
    expect(route.ACTION_HANDLERS.TEST_FETCH_SUCCESS).toEqual(AbstractRoute.fetchSuccess);
    expect(route.ACTION_HANDLERS.TEST_FETCH_FAILURE).toEqual(AbstractRoute.fetchFail);
  });

  it('Renders correctly', () => {
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Has a route', () => {
    expect(connectedRoute).toBeDefined();
    expect(connectedRoute.key).toEqual('test');
    expect(connectedRoute.icon).toEqual('test');
    expect(connectedRoute.path).toEqual('/test');
  });

  it('Can load', () => {
    const dumbComponent = component.root.findByType(route.component);
    dumbComponent.props.load();
  });
});
