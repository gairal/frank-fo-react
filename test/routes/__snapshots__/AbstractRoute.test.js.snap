// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routes > AbstractRoute Renders correctly 1`] = `<div />`;
